File: app/src/main/java/com/github/chuross/recyclerview/sample/MainActivity.java
package com.github.chuross.recyclerview.sample;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.github.chuross.recyclerviewadapters.CompositeRecyclerAdapter;
import com.github.chuross.recyclerviewadapters.ItemAdapter;
import com.github.chuross.recyclerviewadapters.OnItemClickListener;
import com.github.chuross.recyclerviewadapters.OnItemDoubleClickListener;
import com.github.chuross.recyclerviewadapters.OnItemLongPressedListener;
import com.github.chuross.recyclerviewadapters.SpanSizeLookupBuilder;
import com.github.chuross.recyclerviewadapters.ViewItem;
import com.github.chuross.recyclerviewadapters.ViewItemAdapter;


public class MainActivity extends AppCompatActivity {

    private static final int SPAN_SIZE = 4;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final CompositeRecyclerAdapter compositeAdapter = new CompositeRecyclerAdapter();

        final ItemAdapter<String, RecyclerView.ViewHolder> itemAdapter1 = new ItemAdapter<String, RecyclerView.ViewHolder>(this) {
            @Override
            public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int itemViewType) {
                LayoutInflater inflater = LayoutInflater.from(parent.getContext());
                return new RecyclerView.ViewHolder(inflater.inflate(R.layout.item_adapter_1, parent, false)) {
                };
            }

            @Override
            public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
                ((TextView) holder.itemView.findViewById(R.id.text)).setText(get(position));
            }

            @Override
            public int getAdapterId() {
                return R.layout.item_adapter_1;
            }
        };
        itemAdapter1.add("itemAdapter1#default");
        itemAdapter1.setOnItemClickListener(new OnItemClickListener<String>() {
            @Override
            public void onItemClicked(@NonNull RecyclerView.ViewHolder holder, int position, @NonNull String item) {
                Toast.makeText(holder.itemView.getContext(), "click! adapter1:: " + item, Toast.LENGTH_SHORT).show();
            }
        });
        itemAdapter1.setOnItemDoubleClickListener(new OnItemDoubleClickListener<String>() {
            @Override
            public void onItemDoubleClicked(@NonNull RecyclerView.ViewHolder holder, int position, @NonNull String item) {
                Toast.makeText(holder.itemView.getContext(), "double click! adapter1:: " + item, Toast.LENGTH_SHORT).show();
            }
        });
        itemAdapter1.setOnItemLongPressListener(new OnItemLongPressedListener<String>() {
            @Override
            public void onItemLongPressed(@NonNull RecyclerView.ViewHolder holder, int position, @NonNull String item) {
                Toast.makeText(holder.itemView.getContext(), "long press! adapter1:: " + item, Toast.LENGTH_SHORT).show();
            }
        });

        final ItemAdapter<String, RecyclerView.ViewHolder> itemAdapter2 = new ItemAdapter<String, RecyclerView.ViewHolder>(this) {
            @Override
            public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int itemViewType) {
                LayoutInflater inflater = LayoutInflater.from(parent.getContext());
                return new RecyclerView.ViewHolder(inflater.inflate(R.layout.item_adapter_2, parent, false)) {
                };
            }

            @Override
            public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
                ((TextView) holder.itemView.findViewById(R.id.text)).setText(get(position));
            }

            @Override
            public int getAdapterId() {
                return R.layout.item_adapter_2;
            }
        };
        itemAdapter2.add("itemAdapter2#default");
        itemAdapter2.setOnItemClickListener(new OnItemClickListener<String>() {
            @Override
            public void onItemClicked(@NonNull RecyclerView.ViewHolder holder, int position, @NonNull String item) {
                Toast.makeText(holder.itemView.getContext(), "click! adapter2:: " + item, Toast.LENGTH_SHORT).show();
            }
        });

        // same as itemAdapter1 layout resource
        final ItemAdapter<String, RecyclerView.ViewHolder> itemAdapter3 = new ItemAdapter<String, RecyclerView.ViewHolder>(this) {
            @Override
            public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int itemViewType) {
                LayoutInflater inflater = LayoutInflater.from(parent.getContext());
                return new RecyclerView.ViewHolder(inflater.inflate(R.layout.item_adapter_1, parent, false)) {
                };
            }

            @Override
            public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
                ((TextView) holder.itemView.findViewById(R.id.text)).setText(get(position));
            }

            // Replace adapterId, if you want to same layout ItemAdapter.
            @Override
            public int getAdapterId() {
                return R.id.extra_item_adapter;
            }
        };
        itemAdapter3.add("itemAdapter3# same layout as itemAdapter1");
        itemAdapter3.add("itemAdapter3# same layout as itemAdapter1");
        itemAdapter3.add("itemAdapter3# same layout as itemAdapter1");

        ViewItem viewItem1 = new ViewItem(this, R.layout.item_hello_world);
        ViewItem viewItem2 = new ViewItem(this, R.layout.item_append_buttom, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                itemAdapter1.add("itemAdapter1#" + String.valueOf(itemAdapter1.getItemCount()));
                itemAdapter1.add("itemAdapter1#" + String.valueOf(itemAdapter1.getItemCount()));
            }
        });
        ViewItem viewItem3 = new ViewItem(this, R.layout.item_append_buttom, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                itemAdapter2.add("itemAdapter2#" + String.valueOf(itemAdapter2.getItemCount()));
            }
        });

        final ViewItemAdapter viewItemAdapter = new ViewItemAdapter(this, R.id.recyclerviewadapters_itemview_adapter);

        final ViewItem viewItem4 = new ViewItem(this, R.layout.item_footer_1);
        final ViewItem viewItem5 = new ViewItem(this, R.layout.item_footer_2);
        final ViewItem viewItem6 = new ViewItem(this, R.layout.item_append_buttom, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                viewItemAdapter.addAll(
                        viewItem4.clone(),
                        viewItem5.clone()
                );
            }
        });

        viewItemAdapter.add(viewItem4);
        viewItemAdapter.add(viewItem5);

        /*
        compositeAdapter.addAll(
            viewItem1,
            itemAdapter1,
            viewItem2,
            ...
        );
         */
        compositeAdapter.add(viewItem1);
        compositeAdapter.add(itemAdapter1);
        compositeAdapter.add(viewItem2);
        compositeAdapter.add(itemAdapter2);
        compositeAdapter.add(viewItem3);
        compositeAdapter.add(itemAdapter3);
        compositeAdapter.add(viewItemAdapter);
        compositeAdapter.add(viewItem6);

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.list);

        /*
         * simple LinearLayoutManager example
         */
        // recyclerView.setLayoutManager(new LinearLayoutManager(this));


        /*
         * Use SpanSizeLookupBuilder, if you want to use GridLayoutManager with SpanSizeLookup.
         */
        GridLayoutManager gridLayoutManager = new GridLayoutManager(this, SPAN_SIZE);
        recyclerView.setLayoutManager(gridLayoutManager);
        gridLayoutManager.setSpanSizeLookup(new SpanSizeLookupBuilder(this, compositeAdapter)
                .bind(viewItem1, SPAN_SIZE)
                .bind(viewItem2, SPAN_SIZE)
                .bind(viewItem3, SPAN_SIZE)
                .bind(viewItem6, SPAN_SIZE)
                .bind(viewItemAdapter, SPAN_SIZE)
                .bind(itemAdapter1, 2, 1) //can set spanSize separately.
                .bind(itemAdapter3, SPAN_SIZE)
                .build());

        recyclerView.setAdapter(compositeAdapter);
    }
}

File: app/src/main/kotlin/com/github/chuross/recyclerview/sample/MainActivity.kt
package com.github.chuross.recyclerview.sample

import android.databinding.DataBindingUtil
import android.graphics.Color
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.View
import android.widget.Toast
import com.github.chuross.recyclerview.sample.databinding.ActivityMainBinding
import com.github.chuross.recyclerviewadapters.*
import io.reactivex.processors.BehaviorProcessor

class MainActivity : AppCompatActivity() {

    private val maxSpanSize = 4

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val binding = DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)

        val compositeRecyclerAdapter = CompositeRecyclerAdapter()

        val headerViewItem = ViewItem(this, R.layout.item_hello_world)
        val headerVisibilityToggleButton = ViewItem(this, R.layout.visible_toggle, View.OnClickListener {
            headerViewItem.isVisible = !headerViewItem.isVisible
        })

        val itemAdapter1 = TextItemAdapter(this).also {
            it.add("itemAdapter1#0")
            it.setOnItemClickListener { _, _, item ->
                Toast.makeText(this, "click! adapter1::$item", Toast.LENGTH_SHORT).show()
            }
            it.setOnItemLongPressListener { _, _, item ->
                Toast.makeText(this, "long press! adapter1::$item", Toast.LENGTH_SHORT).show()
            }
        }
        val itemAdapter1AppendButton = AppendButtonViewItem(this, View.OnClickListener {
            itemAdapter1.add("itemAdapter1#" + itemAdapter1.itemCount)
            itemAdapter1.add("itemAdapter1#" + itemAdapter1.itemCount)
        })

        // RxJava#Flowable binding adapter
        val itemAdapter2RxSource = BehaviorProcessor.createDefault(listOf("itemAdapter2#0"))
        val itemAdapter2 = GridTextItemAdapter(this, itemAdapter2RxSource).also {
            it.setOnItemClickListener { _, _, item ->
                Toast.makeText(this, "click! adapter2::$item", Toast.LENGTH_SHORT).show()
            }
        }
        val itemAdapter2AppendButton = AppendButtonViewItem(this, View.OnClickListener {
            val value = "RxItemAdapter2#" + itemAdapter2.itemCount.toString()
            itemAdapter2RxSource.onNext(itemAdapter2RxSource.value.plus(value))
        })
        val itemAdapter2VisibilityToggleButton = ViewItem(this, R.layout.visible_toggle, View.OnClickListener {
            itemAdapter2.setVisible(!itemAdapter2.isVisible, true)
        })

        val itemAdapter3 = TextItemAdapter(this, Color.MAGENTA).also {
            it.add("itemAdapter3# same as itemAdapter1's layout [draggable] #1")
            it.add("itemAdapter3# same as itemAdapter1's layout [draggable] #2")
            it.add("itemAdapter3# same as itemAdapter1's layout [draggable] #3")
        }

        val nestedTextAdapter = TextItemAdapter(this).also {
            it.add("nested item adapter #1-1")
            it.add("nested item adapter #1-1")
        }
        val deepNestedTextAdapter = TextItemAdapter(this, Color.parseColor("#388E3C")).also {
            it.add("deep nested item [draggable] #2-1")
            it.add("deep nested item [draggable] #2-2")
        }
        val nestedAdapter = CompositeRecyclerAdapter().also {
            it.add(nestedTextAdapter)
            it.add(CompositeRecyclerAdapter().also {
                it.add(deepNestedTextAdapter)
            })
        }

        val viewItemAdapter = ViewItemAdapter(this).also {
            it.add(ViewItem(this, R.layout.item_footer_1))
            it.add(ViewItem(this, R.layout.item_footer_2))
        }
        val viewItemAdapterAppendButton = AppendButtonViewItem(this, View.OnClickListener {
            viewItemAdapter.add(ViewItem(this, R.layout.item_footer_1))
        })

        /**
         * add to CompositeRecyclerAdapter
         * will be merged all LocalAdapter!
         */
        compositeRecyclerAdapter.addAll(
                headerViewItem,
                headerVisibilityToggleButton,
                itemAdapter1,
                itemAdapter1AppendButton,
                itemAdapter2,
                itemAdapter2AppendButton,
                itemAdapter2VisibilityToggleButton,
                itemAdapter3,
                nestedAdapter,
                viewItemAdapter,
                viewItemAdapterAppendButton
        )


        /**
         * setup RecyclerView by CompositeRecyclerAdapter.
         */
        val recyclerView = binding.list
        val gridLayoutManager = GridLayoutManager(this, maxSpanSize)

        recyclerView.also {
            // can use LinearLayoutManager
            it.layoutManager = gridLayoutManager
            it.adapter = compositeRecyclerAdapter
        }


        /**
         * SpanSizeLookup support
         * @warning must be use GridLayoutManager!!
         */
        gridLayoutManager.spanSizeLookup = SpanSizeLookupBuilder(compositeRecyclerAdapter)
                .register(headerViewItem, maxSpanSize)
                .register(headerVisibilityToggleButton, maxSpanSize)
                .register(itemAdapter1, maxSpanSize)
                .register(itemAdapter2VisibilityToggleButton, maxSpanSize)
                .register(itemAdapter3, maxSpanSize)
                .register(nestedTextAdapter, maxSpanSize)
                .register(deepNestedTextAdapter, maxSpanSize)
                .register(viewItemAdapter, maxSpanSize)
                .register(AppendButtonViewItem::class.java, maxSpanSize) //can also register Class
                .build()


        /**
         * GridPadding support
         * @warning must be use GridLayoutManager!!
         */
        val padding = 16
        recyclerView.addItemDecoration(GridPaddingItemDecorationBuilder(compositeRecyclerAdapter, padding, maxSpanSize)
                .paddingType(GridPaddingItemDecorationBuilder.PaddingType.BOTH)
                .register(headerVisibilityToggleButton)
                .register(itemAdapter2)
                .register(AppendButtonViewItem::class.java)
                .build())

        val nestedItemPadding = 64
        recyclerView.addItemDecoration(GridPaddingItemDecorationBuilder(compositeRecyclerAdapter, nestedItemPadding, maxSpanSize)
                .paddingType(GridPaddingItemDecorationBuilder.PaddingType.VERTICAL)
                .register(deepNestedTextAdapter)
                .build())


        /**
         * List item divider support
         */
        val dividerHeight = resources.getDimensionPixelSize(R.dimen.divider_height) // 1dp

        recyclerView.addItemDecoration(DividerItemDecorationBuilder(compositeRecyclerAdapter)
                .dividerHeight(dividerHeight)
                .dividerColor(Color.BLACK)
                .register(headerVisibilityToggleButton)
                .register(itemAdapter1)
                .register(itemAdapter1AppendButton)
                .register(itemAdapter2AppendButton)
                .register(nestedTextAdapter)
                .register(deepNestedTextAdapter)
                .build())

        recyclerView.addItemDecoration(DividerItemDecorationBuilder(compositeRecyclerAdapter)
                .dividerHeight(dividerHeight)
                .dividerColor(Color.MAGENTA)
                .register(itemAdapter3)
                .build())

        recyclerView.addItemDecoration(DividerItemDecorationBuilder(compositeRecyclerAdapter)
                .dividerHeight(dividerHeight)
                .dividerColor(Color.parseColor("#388E3C"))
                .register(deepNestedTextAdapter)
                .build())


        /**
         * Dragging support
         */
        val dragHelper = DragItemTouchHelperBuilder(compositeRecyclerAdapter)
                .dragFlag(ItemTouchHelper.UP)
                .dragFlag(ItemTouchHelper.DOWN)
                .register(itemAdapter3)
                .register(deepNestedTextAdapter)
                .build()
        dragHelper.attachToRecyclerView(recyclerView)
    }

}